frontendActions/src/main/java/com/virtuslab/gitmachete/frontend/actions/common/PullBackgroundable.java

```diff
@@ -0,0 +1,214 @@
+import com.intellij.openapi.vcs.update.UpdatedFiles;
+import com.intellij.openapi.vfs.VfsUtil;
+import com.intellij.openapi.vfs.VirtualFile;
+import com.intellij.ui.GuiUtils;
+import com.intellij.vcs.ViewUpdateInfoNotification;
+import git4idea.GitBranch;
+import git4idea.GitRevisionNumber;
+import git4idea.GitVcs;
+import git4idea.branch.GitBranchPair;
+import git4idea.commands.Git;
+import git4idea.commands.GitCommandResult;
+import git4idea.commands.GitLineHandler;
+import git4idea.commands.GitLocalChangesWouldBeOverwrittenDetector;
+import git4idea.commands.GitUntrackedFilesOverwrittenByOperationDetector;
+import git4idea.merge.MergeChangeCollector;
+import git4idea.repo.GitRepository;
+import git4idea.update.GitUpdateInfoAsLog;
+import git4idea.update.GitUpdatedRanges;
+import git4idea.util.GitUIUtil;
+import git4idea.util.GitUntrackedFilesHelper;
+import git4idea.util.LocalChangesWouldBeOverwrittenHelper;
+import lombok.CustomLog;
+import org.checkerframework.checker.nullness.qual.Nullable;
+
+@CustomLog
+public class PullBackgroundable extends Task.Backgroundable {
+
+  private static final String OPERATION_NAME = "Pull";
+
+  private final Project project;
+  private final GitRepository gitRepository;
+  private final GitLineHandler handler;
+  private final String remoteBranchName;
+
+  public PullBackgroundable(
+      Project project,
+      GitRepository gitRepository,
+      GitLineHandler handler,
+      String remoteBranchName) {
+    super(project,
+        /* taskTitle */ getString("action.GitMachete.BasePullBranchAction.task-title"),
+        /* canBeCancelled */ true);
+    this.project = project;
+    this.gitRepository = gitRepository;
+    this.handler = handler;
+    this.remoteBranchName = remoteBranchName;
+  }
+
+  @Override
+  public void run(ProgressIndicator indicator) {
+    var localChangesDetector = new GitLocalChangesWouldBeOverwrittenDetector(gitRepository.getRoot(), MERGE);
+    var untrackedFilesDetector = new GitUntrackedFilesOverwrittenByOperationDetector(gitRepository.getRoot());
+
+    handler.addLineListener(localChangesDetector);
+    handler.addLineListener(untrackedFilesDetector);
+
+    Label beforeLabel = LocalHistory.getInstance().putSystemLabel(project, /* name */ "Before update");
+
+    GitUpdatedRanges updatedRanges = deriveGitUpdatedRanges(project, gitRepository, remoteBranchName);
+
+    String beforeRevision = gitRepository.getCurrentRevision();
+    try (AccessToken ignore = DvcsUtil.workingTreeChangeStarted(project, OPERATION_NAME)) {
+      GitCommandResult result = Git.getInstance().runCommand(() -> handler);
+
+      if (beforeRevision != null) {
+        GitRevisionNumber currentRev = new GitRevisionNumber(beforeRevision);
+        handleResult(result,
+            project,
+            localChangesDetector,
+            untrackedFilesDetector,
+            gitRepository,
+            currentRev,
+            beforeLabel,
+            updatedRanges);
+      }
+    }
+  }
+
+  @Nullable
```

Comment:
Nit: since `@Nullable` refers to the return type and not to the method as such: `private static @Nullable GitUpdatedRanges ...`